### PR TITLE
Octahedral Mapping Tests

### PR DESCRIPTION
**Pull request by OmarShehata**
This adds a test for the `OctahedralProjectedCubeMap.js` class. I also removed the temporary +Y/-Y swap from the KTX loader. 

For the sample environment map, can we use something that doesn't have a visible sun, maybe something [like this](https://hdrihaven.com/hdri/?c=outdoor&h=waterbuck_trail) ?

Speaking of which, @bagnell can you write down some steps for generating these KTX files and the spherical harmonics? It might be useful in putting together a tutorial so users know how to use this until there are tools that can export this extension.

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @OmarShehata!
* :heavy_check_mark: Signed CLA found.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

---

__I am a bot who helps you make Cesium awesome! [Contributions to my configuration are welcome.](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/.concierge)__

:earth_africa: :earth_americas: :earth_asia:

**Comment by bagnell** [comment]
Can you add more tests? For example, check that a bunch of samples from the octahedral map are close to samples from the original cube map. Also, do the same for the mipmaps.

**Comment by OmarShehata** [comment]
@bagnell I added a test to sample all 6 directions, for every mip level, and confirm that the octahedral map is pretty close to the original cubemap color. I had to flip the Y to account for the +Y/-Y face swap.

I also had to add a `contextRenderAndReadPixels` function since I couldn't see a way to get the rendered color from a context. I was trying to refactor `expectContextToRender` to use this new function instead of having this duplicated code, but the fact that it checks the cleared color in addition to checking the rendered color makes it a little hard to refactor as-is. I feel like this shouldn't be making these two checks in the same function.

It also seems to be using `CesiumMath.equalsEpsilon` on each individual color channel as opposed to the more general `toEqualEpsilon`, so I'm guessing this is an older function?

File: Specs/Scene/OctahedralProjectedCubeMapSpec.js [hunk 1]
```diff
@@ -1,39 +1,165 @@
...
         }).then(function() {
-            expect(octahedralMap.texture.width).toEqual(194);
-            expect(octahedralMap.texture.height).toEqual(128);
+            expect(octahedralMap.texture.width).toEqual(770);
+            expect(octahedralMap.texture.height).toEqual(512);
             expect(octahedralMap.maximumMipmapLevel).toEqual(5);
         });
     });
+
+    fit('correctly projects the given cube map and all mip levels', function() {
```
Comment (bagnell) [inline Specs/Scene/OctahedralProjectedCubeMapSpec.js]: This should be `it` instead of `fit`.

File: Specs/Scene/OctahedralProjectedCubeMapSpec.js [hunk 2]
```diff
@@ -1,39 +1,165 @@
...
+    }
+
+    it('creates a packed texture with the right dimensions', function() {
         var octahedralMap = new OctahedralProjectedCubeMap(environmentMapUrl);
-        scene.primitives.add(octahedralMap);
+        var frameState = createFrameState(context);
 
         return pollToPromise(function() {
-            scene.renderForSpecs();
+            octahedralMap.update(frameState);
```
Comment (bagnell) [inline Specs/Scene/OctahedralProjectedCubeMapSpec.js]: Use the scene method here instead.

File: Specs/Scene/OctahedralProjectedCubeMapSpec.js [hunk 3]
```diff
@@ -1,39 +1,165 @@
...
         });
     });
+
+    fit('correctly projects the given cube map and all mip levels', function() {
+        var octahedralMap = new OctahedralProjectedCubeMap(environmentMapUrl);
+        var frameState = createFrameState(context);
+
+        return pollToPromise(function() {
+            octahedralMap.update(frameState);
+            executeCommands(frameState);
```
Comment (bagnell) [inline Specs/Scene/OctahedralProjectedCubeMapSpec.js]: Also use the scene here. You can get `scene.context`  for the rest of the test.

**Comment by bagnell** [comment]
> but the fact that it checks the cleared color in addition to checking the rendered color makes it a little hard to refactor as-is. I feel like this shouldn't be making these two checks in the same function.

Why not? The idea is that the framebuffer should be cleared to check that the command executed actual renders the right color. If it isn't cleared, the color might be the same from a previous render.

**Comment by OmarShehata** [comment]
> Why not? The idea is that the framebuffer should be cleared to check that the command executed actual renders the right color. If it isn't cleared, the color might be the same from a previous render.

Wouldn't it work just as well if that was a single separate test to check that the clear command works, instead of every color test also checking it?

That would allow me to refactor `expectContextToRender` to just use `contextRenderAndReadPixels`, and the latter would always return a color. And then add a separate test somewhere that clear works.

**Comment by bagnell** [comment]
> I was trying to refactor `expectContextToRender` to use this new function instead of having this duplicated code

Could you also do this? You could have a function return the clear color and render color. If the test isn't supposed to clear first, then don't clear and return undefined. You're test should also clear before rendering the cube map face and before the sampling the octahedral projection.

**Comment by bagnell** [comment]
> Wouldn't it work just as well if that was a single separate test to check that the clear command works, instead of every color test also checking it?

I don't understand. There is a test to check that a clear works. The clear before render is to prevent something like this:
```javascript
it('test 1', function() {
    // set up
    expect(scene).toRender(color);
});
it('test 2', function() {
    // set up
    expect(scene).toRender(color);
});
```
Now, say test 1 runs first and renders `color`. Now when test 2 runs and doesn't render anything. If the color wasn't cleared, it will pass and that's a bug.

**Comment by OmarShehata** [comment]
> Now, say test 1 runs first and renders color. Now when test 2 runs and doesn't render anything. If the color wasn't cleared, it will pass and that's a bug.

In that scenario, test 2 would pass, but the "check that a clear works" would fail, right?

**Comment by bagnell** [comment]
> but the "check that a clear works" would fail, right?

In that scenario, say the clear test runs before or after both of those tests execute. Now all of the tests still pass when test 2 should fail.

**Comment by bagnell** [comment]
No, the scenario was to show what could happen if there is no clear before every render.

**Comment by OmarShehata** [comment]
So yes, we're still going to clear before each test, it's just that testing that it correctly clears in every test is redundant, but if this redundancy helps here then I'm all for it. Just wanted to make sure we were on the same page. 

@bagnell fixed the `fit`, explained why we have to execute the commands in this test as opposed to using `scene` methods, and I refactor the matchers. It now returns the rendered color as well as the clear color. Everything should work as before, and I confirmed before/after that the same tests are failing (the same 10 failures that are already in this branch).

**Pull request merged**